File: Java/src/com/leetcode/Medium/MinimumAddToMakeParenthesesValid.java
package com.leetcode.Medium;

import java.util.Stack;

// 921. Minimum Add to Make Parentheses Valid
public class MinimumAddToMakeParenthesesValid {
    public int minAddToMakeValid(String S) {
        Stack<Character> stack = new Stack<>();
        for (char ch : S.toCharArray()) {
            if (!stack.isEmpty()) {
                if (isMatchingPair(stack.peek(), ch))
                    stack.pop();
                else
                    stack.push(ch);
            } else {
                stack.push(ch);
            }
        }
        return stack.size();
    }

    private static boolean isMatchingPair(char character1, char character2) {
        return character1 == '(' && character2 == ')';
    }
}

File: JavaScript/Medium/MinimumAddToMakeParenthesesValid.js
/**
 * 921. Minimum Add to Make Parentheses Valid
 * @param {string} s
 * @return {number}
 */
const minAddToMakeValid = function (s) {
  let stack = new Array();
  for (let ch of s) {
    if (stack.length > 0) {
      if (isMatchingPair(stack[stack.length - 1], ch))
        stack.pop();
      else
        stack.push(ch);
    } else {
      stack.push(ch);
    }
  }
  return stack.length;
};

const isMatchingPair = function (ch1, ch2) {
  return ch1 == '(' && ch2 == ')';
}